File: java/invaders_game/Barrier.java
/**
 * ファイル名：Barrier.java
 * 説明：バリア1基のセル集合・弾衝突による段階的破壊・描画を担当する。
 * 作成日：2026-04-10
 * 作者：takashioikawa
 */

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * バリア1基。小さな矩形セルの集合として保持し、
 * 弾が当たったセルから順に欠けていく（段階的破壊）。
 */
public class Barrier {

    private final List<Rectangle> blocks = new ArrayList<>();

    /**
     * @param centerX バリア全体の中心X
     * @param topY    バリア上端Y
     */
    public Barrier(int centerX, int topY) {
        int cols = 11;
        int rows = 5;
        int cellW = 7;
        int cellH = 6;
        int totalW = cols * cellW;
        int left = centerX - totalW / 2;
        // 下部中央をくり抜いたクラシックに近い形状
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (r == rows - 1 && c >= 2 && c <= cols - 3) {
                    continue;
                }
                blocks.add(new Rectangle(left + c * cellW, topY + r * cellH, cellW - 1, cellH - 1));
            }
        }
    }

    /**
     * 弾との衝突を検査する。当たったセルを1つ削除し true を返す。
     *
     * @param bulletRect 弾の矩形
     * @return 何かしらのセルに当たったか
     */
    public boolean collideBullet(Rectangle bulletRect) {
        Iterator<Rectangle> it = blocks.iterator();
        while (it.hasNext()) {
            Rectangle block = it.next();
            if (block.intersects(bulletRect)) {
                it.remove();
                return true;
            }
        }
        return false;
    }

    /** 残っているセルを描画する。 */
    public void draw(Graphics2D g2) {
        for (Rectangle block : blocks) {
            g2.setColor(new Color(90, 220, 120));
            g2.fillRect(block.x, block.y, block.width, block.height);
            g2.setColor(new Color(20, 80, 30));
            g2.drawRect(block.x, block.y, block.width, block.height);
        }
    }
}

File: python/invaders_game/barrier.py
"""シールド（バリア）の段階的破壊を担当するモジュール。"""

from typing import List

import pygame


class Barrier:
    """
    バリア1基。小さな矩形セルの集合として保持し、
    弾が当たったセルから順に欠けていく（段階的破壊）。
    """

    def __init__(self, center_x: int, top_y: int) -> None:
        """
        Args:
            center_x: バリア全体の中心X
            top_y: バリア上端Y
        """
        self.blocks: List[pygame.Rect] = []
        cols, rows = 11, 5
        cell_w, cell_h = 7, 6
        total_w = cols * cell_w
        left = center_x - total_w // 2
        # 下部中央をくり抜いたクラシックに近い形状
        for r in range(rows):
            for c in range(cols):
                if r == rows - 1 and 2 <= c <= cols - 3:
                    continue
                rect = pygame.Rect(left + c * cell_w, top_y + r * cell_h, cell_w - 1, cell_h - 1)
                self.blocks.append(rect)

    def collide_bullet(self, bullet_rect: pygame.Rect) -> bool:
        """
        弾との衝突を検査する。当たったセルを1つ削除しTrueを返す。
        """
        for i, block in enumerate(self.blocks):
            if block.colliderect(bullet_rect):
                del self.blocks[i]
                return True
        return False

    def draw(self, surface: pygame.Surface) -> None:
        """残っているセルを描画する。"""
        for block in self.blocks:
            pygame.draw.rect(surface, (90, 220, 120), block)
            pygame.draw.rect(surface, (20, 80, 30), block, width=1)
